packages/monitor-v2/src/monitor-og/SnapshotVerification.ts

```diff
@@ -0,0 +1,342 @@
+  } catch (error) {
+    return new Error(JSON.stringify(error));
+  }
+};
+
+// We don't want to throw an error if the IPFS request fails for any reason, so we return a stringified Error object
+// instead that will be logged by the bot.
+const getIpfsData = async (ipfsHash: string, url: string): Promise<IpfsData | Error> => {
+  try {
+    const response = await fetch(`${url}/${ipfsHash}`);
+    if (!response.ok) {
+      return new Error(`Request on ${response.url} failed with status ${response.status}`);
+    }
+    const data = await response.json();
+
+    // Try to parse the plugins property as JSON.
+    data.data.message.plugins = JSON.parse(data.data.message.plugins);
+    return data;
+  } catch (error) {
+    return new Error(JSON.stringify(error));
+  }
+};
+
+const ipfsMatchGraphql = (ipfsData: IpfsData, graphqlData: GraphqlData): boolean => {
+  const ipfsProposal = ipfsData.data.message;
+  const graphqlProposal = graphqlData.proposals[0];
+
+  // Verify common properties, except for safeSnap plugin.
+  if (
+    ipfsProposal.space !== graphqlProposal.space.id ||
+    ipfsProposal.type !== graphqlProposal.type ||
+    JSON.stringify(ipfsProposal.choices) !== JSON.stringify(graphqlProposal.choices) ||
+    ipfsProposal.start !== graphqlProposal.start ||
+    ipfsProposal.end !== graphqlProposal.end
+  ) {
+    return false;
+  }
+
+  // Verify that the safeSnap plugin exists on both data sources and has the same number of safes.
+  if (
+    ipfsProposal.plugins.safeSnap === undefined ||
+    graphqlProposal.plugins.safeSnap === undefined ||
+    ipfsProposal.plugins.safeSnap.safes.length !== graphqlProposal.plugins.safeSnap.safes.length
+  ) {
+    return false;
+  }
+
+  // Verify that the safes match. We inspect only the properties that are required for verification.
+  for (let i = 0; i < ipfsProposal.plugins.safeSnap.safes.length; i++) {
+    const ipfsSafe = ipfsProposal.plugins.safeSnap.safes[i];
+    const graphqlSafe = graphqlProposal.plugins.safeSnap.safes[i];
+
+    // Verify the network, oSnap address and number of transaction batches.
+    if (
+      ipfsSafe.network !== graphqlSafe.network ||
+      ipfsSafe.umaAddress !== graphqlSafe.umaAddress ||
+      ipfsSafe.txs.length !== graphqlSafe.txs.length
+    ) {
+      return false;
+    }
+
+    // Verify that the transaction batch properties match.
+    for (let j = 0; j < ipfsSafe.txs.length; j++) {
+      const ipfsTransaction = ipfsSafe.txs[j].mainTransaction;
+      const graphqlTransaction = graphqlSafe.txs[j].mainTransaction;
+
+      if (
+        ipfsTransaction.to !== graphqlTransaction.to ||
+        ipfsTransaction.data !== graphqlTransaction.data ||
+        ipfsTransaction.value !== graphqlTransaction.value ||
+        ipfsTransaction.operation !== graphqlTransaction.operation
+      ) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+};
+
+const findSafe = (safe: SafeSnapSafe, chainId: number, ogAddress: string): boolean => {
```

Comment:
nit: naming, I think a more accurate name would be `isMatchingSafe` or similar since this doesn't find the safe, it just matches it.